typeahead: Use custom profile fields in user mention suggestions.

Asteriskkkk opened the pull request:
<!-- Describe your pull request here.-->
This PR adds support for matching selected custom profile fields in the @-mention typeahead.

Key changes:

- Enables matching on:
  - **External account** fields (default: on)
  - **Text (short)** fields (default: off)
  
1. Adds a new setting on External Account and Text (short) custom profile field: “Use in @-mention suggestions.”
2. When a exact match comes from a custom profile field, the field value is shown as secondary text.
3. Falls back to showing the delivery email when no custom field match applies.
4. If no custom profile field match occurs, the typeahead **continues showing the delivery email** as before.

CZO #design thread: [#design > use custom profile fields in user typeahead #36458](https://chat.zulip.org/#narrow/channel/101-design/topic/use.20custom.20profile.20fields.20in.20user.20typeahead.20.2336458/with/2291061)
CZO #api-design thread: [#api design > Use custom profile fields in user mention suggestions.](https://chat.zulip.org/#narrow/channel/378-api-design/topic/Use.20custom.20profile.20fields.20in.20user.20mention.20suggestions.2E/with/2312589)

Fixes: [36458](https://github.com/zulip/zulip/issues/36458)

<details>
<summary>Screenshots and screen captures</summary>

### Use in @-mention checkbox: Text (short) default (off), External Account default(on)  

| Before | After |
| ------ | ------ |
| ![image-before](https://github.com/user-attachments/assets/41a3bb37-74d6-4dad-87f4-070118d1d0c3) | ![image-after](https://github.com/user-attachments/assets/14c40ee8-d8c9-47f6-bd42-6875902ee891)
| ![image-before](https://github.com/user-attachments/assets/f784790e-3a0e-4fdf-88fb-c98a2a60d0c1)| ![image-after](https://github.com/user-attachments/assets/4cde9030-00ed-4d89-8b2c-b02f5ac28e0a) 


### When custom field (text (short), external account) matches exactly with user query
| Before | After |
| ------ | ------ |
| ![image-before](https://github.com/user-attachments/assets/c283ebc9-f757-4a56-aca0-6857f575e3b4)| ![image-after](https://github.com/user-attachments/assets/904e2976-db2b-43b6-b73a-cf6067fd97b1) 

</details>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! Please go through our [guide on submitting a pull request](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html) to learn how to present your proposed changes to Zulip, so that your work can be reviewed.

Asteriskkkk commented:
> Thanks! Please go through our [guide on submitting a pull request](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html) to learn how to present your proposed changes to Zulip, so that your work can be reviewed.

Thanks. I’ve reviewed the pull request submission guide and updated my PR accordingly.
Please let me know if there are any further suggestions or adjustments needed.

As I’m a new contributor, I also wanted to confirm whether I should address the failing GitHub Actions checks?.

Asteriskkkk commented:
Updated the PR description as per the feedback. I’ve also added a shorter video and images to demonstrate the changes. Please review when you get a moment. Thanks.

alya commented:
@sahil839 Are you up for doing the first round of review on this one? I haven't tested.

sahil839 commented on zerver/actions/custom_profile_fields.py:
```diff
@@ -57,14 +62,24 @@ def try_add_realm_custom_profile_field(
+        elif field_type == CustomProfileField.SHORT_TEXT:
+            use_in_mentions = False
+        else:
+            use_in_mentions = False
```
You should also add a validate if `use_in_mentions` value is passed for field types other than "External account" and "Short text". I am not sure if we would want to allow passing `False` or not. This can be discussed once on CZO.

sahil839 commented on zerver/views/custom_profile_fields.py:
```diff
@@ -241,6 +244,7 @@ def update_realm_custom_profile_field(
     hint: str | None = None,
     name: Annotated[str, StringConstraints(strip_whitespace=True)] | None = None,
     required: Json[bool] | None = None,
+    use_in_mentions: Json[bool] | None = None,
```
Was this API change discussed about finalizing the name, etc.? If not we should discuss this in #api-design on CZO.

Asteriskkkk replied:
Discussed in the CZO [#api design > Use custom profile fields in user mention suggestions.](https://chat.zulip.org/#narrow/channel/378-api-design/topic/Use.20custom.20profile.20fields.20in.20user.20mention.20suggestions.2E/with/2318002) 
we will be moving forward with `use_for_user_matching`

sahil839 commented on zerver/models/custom_profile_fields.py:
```diff
@@ -174,9 +178,12 @@ def as_dict(self) -> ProfileDataElementBase:
         }
         if self.display_in_profile_summary:
             data_as_dict["display_in_profile_summary"] = True
+        if self.use_in_mentions:
```
This does not make sense since you alread set `use_in_mentions` above.

Would be good to confirm though if we want to include this value in the object if the value is `False`.

sahil839 commented on zerver/tests/test_custom_profile_data.py:
```diff
@@ -694,6 +694,32 @@ def test_update_display_in_profile_summary(self) -> None:
+        )
+        self.assert_json_success(result)
+        field.refresh_from_db()
+        self.assertFalse(field.use_in_mentions)
```
These changes should be in the previous commit which adds the field. And also you should write other tests for setting `use_in_mentions` while creating custom fields and also a test for how the default value is handled.

sahil839 commented on web/templates/settings/add_new_custom_profile_field_form.hbs:
```diff
@@ -62,5 +62,12 @@
+                <span class="rendered-checkbox"></span>
+                {{t 'Use in @-mention suggestions' }}
+            </label>
+        </div>
```
I think this should use `settings_checkbox` template.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -25728,6 +25740,20 @@ components:
+            suggestions.
+
+            This setting is only applicable to profile field types that store text values:
+            **Short text (type 1)** and **External account (type 7)**.
```
I don't think it is needed to mention 1 and 7 values in the type. And maybe we can format this similar to how we format description for `display_in_profile_summary`.

Asteriskkkk commented on web/src/typeahead_helper.ts:
```diff
@@ -1072,6 +1072,28 @@ export function query_matches_person(
+                        " ",
+                    )
+                ) {
+                    return true;
```
should it be implemented inside the `get_person_suggestions()` in`composebox_typeahead.ts` i think its the correct place as it works for only @ mentions
correct me if wrong.

Asteriskkkk commented on web/src/settings_profile_fields.ts:
```diff
@@ -219,6 +219,22 @@ function set_up_create_field_form(): void {
+        );
+    } else {
+        $("#profile_field_use_for_user_matching_row").hide();
+        $("#profile_field_use_for_user_matching").prop("checked", false);
```
Sorry, I missed this in the earlier PR. I’ve updated the PR to address it.

Asteriskkkk commented:
@sahil839 updated the PR! please review when you get a chance.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -1072,6 +1073,30 @@ export function query_matches_person(
+                if (
+                    field.use_for_user_matching &&
+                    (field.type === field_types.SHORT_TEXT.id ||
+                        field.type === field_types.EXTERNAL_ACCOUNT.id)
```
I don't think we need to check `field.type` here as `use_for_user_matching` should be true only for these types.

sahil839 commented on web/src/composebox_typeahead.ts:
```diff
@@ -781,7 +781,13 @@ export function get_person_suggestions(
+                item,
+                should_remove_diacritics,
+                undefined,
+                true,
```
I don't think this correct since now the users with custom profile field matched are shown in DM typeahead as well. I think you would need to add a parameter to `get_person_suggestions` and that should be passed here instead of always passing this as true as `get_person_suggestions` is used for other typeaheads as well apart from mention typeahead.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -112,6 +115,30 @@ export let render_person = (person: UserPillData | UserOrMentionPillData): strin
 
     const pronouns = pronouns_list?.[0]?.value;
 
+    // Prefer showing a matching custom profile field over email when the query matches it.
```
We are showing custom profile field only if email does not matches. I am not sure what is the expected behavior if both match. This comment as of now feels like we would show custom field over email if both matches but that is not the case.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -1072,6 +1073,25 @@ export function query_matches_person(
+                    const field_value =
+                        people.get_custom_profile_data(person.user.user_id, field.id)?.value ?? "";
+                    if (
+                        typeahead.query_matches_string_in_order_assume_canonicalized(
```
Should we use `query_matches_string_in_order` here? I am not sure if we have any restriction of diacritics in custom profile field values.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -112,6 +115,37 @@ export let render_person = (person: UserPillData | UserOrMentionPillData): strin
+
+    const q = query?.toLowerCase();
+    if (q) {
+        const email_matches = secondary_text?.toLowerCase().includes(q) ?? false;
```
Can we use `typeahead.query_matches_string_in_order` functions here and below as well when matching custom field directly instead of having same steps like to convert it to lowercase repeatedly?

sahil839 commented:
Looks good to me. @alya do you want to test this once?

timabbott commented on zerver/migrations/0772_customprofilefield_use_for_user_matching.py:
```diff
@@ -0,0 +1,16 @@
+        migrations.AddField(
+            model_name="customprofilefield",
+            name="use_for_user_matching",
+            field=models.BooleanField(default=False),
```
Set `db_default=False` to make the migration faster.
